extensions/amp-story/1.0/parallax-fx.js

```diff
@@ -0,0 +1,352 @@
+import {toArray} from '../../../src/types';
+
+/** @enum {string} */
+export const ParallaxModes = {
+  /**
+   * Indicates that the last (nearest) layer will appear at the
+   * surface of the screen and all previous layers are behind it
+   * "inside" the screen.
+   */
+  DEPTH: 'depth',
+
+  /**
+   * Indicates that the first (farthest) layer will appear at the surface
+   * of the screen and all subsequent layers will appear as "popping out"
+   * of the screen.
+   */
+  POP_OUT: 'pop-out',
+
+  /**
+   * Indicates that the middle layer will appear at the surface of the
+   * screen with the previous (farther) layers appearing behind the screen
+   * and the subsequent (nearer) layers popping out of the screen.
+   */
+  CENTER: 'center',
+};
+
+const SMOOTHING_PTS = 4;
+const PERSPECTIVE = 1500;
+const DEFAULT_MODE = ParallaxModes.DEPTH;
+const DEFAULT_LAYER_SPACING = 1;
+const DEFAULT_FARTHEST_SCALE = 1.2;
+const DEFAULT_NEAREST_SCALE = 1;
+
+const MODE_ATTR = 'parallax-fx-mode';
+const LAYER_SPACING_ATTR = 'parallax-fx-layer-spacing';
+const NEAREST_SCALE_ATTR = 'parallax-fx-nearest-scale';
+const FARTHEST_SCALE_ATTR = 'parallax-fx-farthest-scale';
+const NO_PARALLAX_FX_ATTR = 'no-parallax-fx';
+
+/**
+ * Installs parallax handlers
+ */
+export class ParallaxService {
+  /**
+   * @param {!Window} global
+   * @param {!../../../src/service/vsync-impl.Vsync} vsync
+   * @param {!Element} story
+   */
+  constructor(global, vsync, story) {
+    /** @private {!Window} */
+    this.win_ = global;
+
+    /** @private @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {!Element} */
+    this.story_ = story;
+
+    /** @private {!Array<!ParallaxPage>} */
+    this.parallaxPages_ = [];
+
+    /** @private {number} */
+    this.middleX_ = 0;
+
+    /** @private {number} */
+    this.middleY_ = 0;
+
+    /** @private {Array} */
+    this.smoothingPointsX_ = [];
+
+    /** @private {Array} */
+    this.smoothingPointsY_ = [];
+
+    /** @private {string} */
+    this.storyMode_ = this.story_.getAttribute(MODE_ATTR);
+
+    /** @private {string} */
+    this.storyLayerSpacing_ = this.story_.getAttribute(LAYER_SPACING_ATTR);
+
+    /** @private {string} */
+    this.storyFarthestScale_ = this.story_.getAttribute(FARTHEST_SCALE_ATTR);
+
+    /** @private {string} */
+    this.storyNearestScale_ = this.story_.getAttribute(NEAREST_SCALE_ATTR);
+
+    this.init_();
+
+    userAssert(
+      this.win_.DeviceOrientationEvent,
+      "The current browser doesn't support the device motion/orientation events"
+    );
+
+    if (this.win_.DeviceOrientationEvent) {
+      listen(this.win_, 'deviceorientation', event => {
+        this.parallaxOrientationMutate_(event, this.parallaxPages_);
+      });
+    }
+  }
+
+  /**
+   * Initializes the pages by giving them perspective and initializating their layers
+   */
+  init_() {
+    this.vsync_.mutate(() => {
```

Comment:
Feels like this logic would make more sense if each page asked the service to do the setup in its `buildCallback` (or perhaps better, its `layoutCallback`). There doesn't seem to be a reason to do work up front for all pages here.